subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c

```diff
@@ -0,0 +1,1251 @@
+		return -ENOMEM;
+	}
+
+	/* Add new endpoint. */
+	ept = &dev_data->ept[ept_index];
+	ept->cfg = cfg;
+	if (dev_data->is_initiator) {
+		ept->addr = ept_index;
+		dev_data->ept_map[ept->addr] = ept->addr;
+	}
+	atomic_set(&ept->state, EPT_CONFIGURED);
+
+	/* Keep endpoint address in token. */
+	*token = ept;
+
+	/* Rest of the bounding will be done in the system workqueue. */
+	schedule_ept_bound_process(dev_data);
+
+	return r;
+}
+
+/**
+ * Returns maximum TX buffer size.
+ */
+static int get_tx_buffer_size(const struct device *instance, void *token)
+{
+	const struct icmsg_with_buf_config *conf = instance->config;
+
+	return conf->tx.block_size * conf->tx.block_count - BLOCK_HEADER_SIZE;
+}
+
+/**
+ * Endpoint TX buffer allocation callback for nocopy sending.
+ */
+static int get_tx_buffer(const struct device *instance, void *token, void **data,
+			 uint32_t *user_len, k_timeout_t wait)
+{
+	struct backend_data *dev_data = instance->data;
+	int r;
+
+	r = alloc_tx_buffer(dev_data, user_len, (uint8_t **)data, wait);
+	if (r < 0) {
+		return r;
+	}
+	return 0;
+}
+
+/**
+ * Endpoint TX buffer release callback for nocopy sending.
+ */
+static int drop_tx_buffer(const struct device *instance, void *token, const void *data)
+{
+	struct backend_data *dev_data = instance->data;
+	int r;
+
+	r =  release_tx_buffer(dev_data, data, -1);
+	if (r < 0) {
+		return r;
+	}
+
+	return 0;
+}
+
+/**
+ * Endpoint nocopy sending.
+ */
+static int send_nocopy(const struct device *instance, void *token, const void *data,
+		       size_t len)
+{
+	struct backend_data *dev_data = instance->data;
+	struct ept_data *ept = token;
+	int r;
+
+	/* Actual size may be smaller than requested, so shrink if possible. */
+	r = release_tx_buffer(dev_data, data, len);
+	if (r < 0) {
+		release_tx_buffer(dev_data, data, -1);
+		return r;
+	}
+
+	return send_block(dev_data, MSG_DATA, ept->addr, r, len);
+}
+
+/**
+ * Holding RX buffer for nocopy receiving.
+ */
+static int hold_rx_buffer(const struct device *instance, void *token, void *data)
+{
+	const struct icmsg_with_buf_config *conf = instance->config;
+	int rx_block_index;
+	uint8_t *buffer = data;
+
+	/* Calculate block index and set associated bit. */
+	rx_block_index = buffer_to_index_validate(&conf->rx, buffer, NULL);
+	__ASSERT_NO_MSG(rx_block_index >= 0);
+	return sys_bitarray_set_bit(conf->rx_hold_bitmap, rx_block_index);
+}
+
+/**
+ * Release RX buffer that was previously held.
+ */
+static int release_rx_buffer(const struct device *instance, void *token, void *data)
+{
+	struct backend_data *dev_data = instance->data;
+
+	return send_release(dev_data, (uint8_t *)data, MSG_RELEASE_DATA, 0);
+}
+
+/**
+ * Backend device initialization.
+ */
+static int backend_init(const struct device *instance)
+{
+	const struct icmsg_with_buf_config *conf = instance->config;
+	struct backend_data *dev_data = instance->data;
+
+	dev_data->conf = conf;
+	dev_data->is_initiator = (conf->rx.blocks_ptr < conf->tx.blocks_ptr);
```

Comment:
Why? Doing it automatically saves a lot of trouble for user, because he don't need to make sure that he put right values for each core. Things gets even more complicated when you have more than one core. What if you want to talk with one core as initiator and with the other not?

Detecting it in runtime makes the bounding a little bit bigger, because program contains support for both cases. I check it when I was implementing it. It was few hundreds bytes (I don't remember exact number now).

In my plans for improvements I have initiator detection in compile time, so during the preprocessing we will know if we should compile code for initiator, follower or both. I didn't implemented it to not over-complicate first version.

Reply:
I was just wonder if this condition will work in feature when maybe different processors could have different memory mapping but I don't sure if such case is even possible. My other question is if we even need to have roles in this module. It seems to me that both sides could send bound message with local endpoint address to each others.